File: spring-boot-project/spring-boot-tools/spring-boot-configuration-processor/src/main/java/org/springframework/boot/configurationprocessor/PropertyDescriptor.java
/*
 * Copyright 2012-2024 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.springframework.boot.configurationprocessor;

import java.util.List;

import javax.lang.model.element.Element;
import javax.lang.model.element.ElementKind;
import javax.lang.model.element.ExecutableElement;
import javax.lang.model.element.TypeElement;
import javax.lang.model.type.TypeMirror;

import org.springframework.boot.configurationprocessor.metadata.ConfigurationMetadata;
import org.springframework.boot.configurationprocessor.metadata.ItemDeprecation;
import org.springframework.boot.configurationprocessor.metadata.ItemMetadata;

/**
 * Description of a property that can be candidate for metadata generation.
 *
 * @author Stephane Nicoll
 * @author Phillip Webb
 */
abstract class PropertyDescriptor {

	private final String name;

	private final TypeMirror type;

	private final TypeElement declaringElement;

	private final ExecutableElement getter;

	/**
	 * Create a new {@link PropertyDescriptor} instance.
	 * @param name the property name
	 * @param type the property type
	 * @param declaringElement the element that declared the item
	 * @param getter the getter for the property or {@code null}
	 */
	PropertyDescriptor(String name, TypeMirror type, TypeElement declaringElement, ExecutableElement getter) {
		this.declaringElement = declaringElement;
		this.name = name;
		this.type = type;
		this.getter = getter;
	}

	/**
	 * Return the name of the property.
	 * @return the property name
	 */
	String getName() {
		return this.name;
	}

	/**
	 * Return the type of the property.
	 * @return the property type
	 */
	TypeMirror getType() {
		return this.type;
	}

	/**
	 * Return the element that declared the property.
	 * @return the declaring element
	 */
	protected final TypeElement getDeclaringElement() {
		return this.declaringElement;
	}

	/**
	 * Return the getter for the property.
	 * @return the getter or {@code null}
	 */
	protected final ExecutableElement getGetter() {
		return this.getter;
	}

	/**
	 * Resolve the {@link ItemMetadata} for this property.
	 * @param prefix the property prefix
	 * @param environment the metadata generation environment
	 * @return the item metadata or {@code null}
	 */
	final ItemMetadata resolveItemMetadata(String prefix, MetadataGenerationEnvironment environment) {
		if (isNested(environment)) {
			return resolveItemMetadataGroup(prefix, environment);
		}
		if (isProperty(environment)) {
			return resolveItemMetadataProperty(prefix, environment);
		}
		return null;
	}

	/**
	 * Return if this is a nested property.
	 * @param environment the metadata generation environment
	 * @return if the property is nested
	 * @see #isMarkedAsNested(MetadataGenerationEnvironment)
	 */
	boolean isNested(MetadataGenerationEnvironment environment) {
		Element typeElement = environment.getTypeUtils().asElement(getType());
		if (!(typeElement instanceof TypeElement) || typeElement.getKind() == ElementKind.ENUM
				|| environment.getConfigurationPropertiesAnnotation(getGetter()) != null) {
			return false;
		}
		if (isMarkedAsNested(environment)) {
			return true;
		}
		return !isCyclePresent(typeElement, getDeclaringElement())
				&& isParentTheSame(environment, typeElement, getDeclaringElement());
	}

	/**
	 * Return if this property has been explicitly marked as nested (for example using an
	 * annotation}.
	 * @param environment the metadata generation environment
	 * @return if the property has been marked as nested
	 */
	protected abstract boolean isMarkedAsNested(MetadataGenerationEnvironment environment);

	private boolean isCyclePresent(Element returnType, Element element) {
		if (!(element.getEnclosingElement() instanceof TypeElement)) {
			return false;
		}
		if (element.getEnclosingElement().equals(returnType)) {
			return true;
		}
		return isCyclePresent(returnType, element.getEnclosingElement());
	}

	private boolean isParentTheSame(MetadataGenerationEnvironment environment, Element returnType,
			TypeElement element) {
		if (returnType == null || element == null) {
			return false;
		}
		returnType = getTopLevelType(returnType);
		Element candidate = element;
		while (candidate instanceof TypeElement) {
			if (returnType.equals(getTopLevelType(candidate))) {
				return true;
			}
			candidate = environment.getTypeUtils().asElement(((TypeElement) candidate).getSuperclass());
		}
		return false;
	}

	private Element getTopLevelType(Element element) {
		if (!(element.getEnclosingElement() instanceof TypeElement)) {
			return element;
		}
		return getTopLevelType(element.getEnclosingElement());
	}

	private ItemMetadata resolveItemMetadataGroup(String prefix, MetadataGenerationEnvironment environment) {
		Element propertyElement = environment.getTypeUtils().asElement(getType());
		String nestedPrefix = ConfigurationMetadata.nestedPrefix(prefix, getName());
		String dataType = environment.getTypeUtils().getQualifiedName(propertyElement);
		String ownerType = environment.getTypeUtils().getQualifiedName(getDeclaringElement());
		String sourceMethod = (getGetter() != null) ? getGetter().toString() : null;
		return ItemMetadata.newGroup(nestedPrefix, dataType, ownerType, sourceMethod);
	}

	private ItemMetadata resolveItemMetadataProperty(String prefix, MetadataGenerationEnvironment environment) {
		String dataType = resolveType(environment);
		String ownerType = environment.getTypeUtils().getQualifiedName(getDeclaringElement());
		String description = resolveDescription(environment);
		Object defaultValue = resolveDefaultValue(environment);
		ItemDeprecation deprecation = resolveItemDeprecation(environment);
		return ItemMetadata.newProperty(prefix, getName(), dataType, ownerType, null, description, defaultValue,
				deprecation);
	}

	private String resolveType(MetadataGenerationEnvironment environment) {
		return environment.getTypeUtils().getType(getDeclaringElement(), getType());
	}

	private ItemDeprecation resolveItemDeprecation(MetadataGenerationEnvironment environment) {
		boolean deprecated = getDeprecatableElements().stream().anyMatch(environment::isDeprecated);
		return deprecated ? environment.resolveItemDeprecation(getGetter()) : null;
	}

	/**
	 * Resolve the property description.
	 * @param environment the metadata generation environment
	 * @return the property description
	 */
	protected abstract String resolveDescription(MetadataGenerationEnvironment environment);

	/**
	 * Resolve the default value for this property.
	 * @param environment the metadata generation environment
	 * @return the default value or {@code null}
	 */
	protected abstract Object resolveDefaultValue(MetadataGenerationEnvironment environment);

	/**
	 * Return all the elements that should be considered when checking for deprecation
	 * annotations.
	 * @return the deprecatable elements
	 */
	protected abstract List<Element> getDeprecatableElements();

	/**
	 * Return true if this descriptor is for a property.
	 * @param environment the metadata generation environment
	 * @return if this is a property
	 */
	abstract boolean isProperty(MetadataGenerationEnvironment environment);

}

File: configuration-metadata/spring-boot-configuration-symbol-processor/src/main/kotlin/org/springframework/boot/configurationprocessor/ksp/PropertyDescriptor.kt
/*
 * Copyright 2012-present the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.springframework.boot.configurationprocessor.ksp

import com.google.devtools.ksp.symbol.ClassKind
import com.google.devtools.ksp.symbol.KSAnnotated
import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.google.devtools.ksp.symbol.KSFunctionDeclaration
import com.google.devtools.ksp.symbol.KSPropertyDeclaration
import com.google.devtools.ksp.symbol.KSType
import com.google.devtools.ksp.symbol.KSValueParameter
import org.springframework.boot.configurationprocessor.ksp.ConfigurationPropertiesAnnotations.CONFIGURATION_PROPERTIES
import org.springframework.boot.configurationprocessor.ksp.ConfigurationPropertiesAnnotations.DEFAULT_VALUE
import org.springframework.boot.configurationprocessor.ksp.ConfigurationPropertiesAnnotations.NESTED_CONFIGURATION_PROPERTY
import org.springframework.boot.configurationprocessor.metadata.ConfigurationMetadata
import org.springframework.boot.configurationprocessor.metadata.ItemMetadata

/**
 * Description of a Kotlin declaration that is a candidate for metadata generation.
 *
 * @author Areg Iazychian
 * @since 4.2.0
 */
internal sealed class PropertyDescriptor(
	val name: String,
	val type: KSType,
	val declaringType: KSClassDeclaration,
) {

	/**
	 * Elements that can carry the annotations describing this property, in the order in
	 * which they should be considered.
	 */
	protected abstract val annotatedElements: List<KSAnnotated>

	/**
	 * Elements that can carry the annotations describing the deprecation of this property.
	 */
	protected open val deprecationElements: List<KSAnnotated>
		get() = annotatedElements

	/**
	 * Return whether this descriptor describes a property rather than a nested group.
	 */
	abstract fun isProperty(context: MetadataGenerationContext): Boolean

	/**
	 * Return the default value of this property, or `null` if it has none or if it cannot
	 * be determined.
	 */
	abstract fun defaultValue(context: MetadataGenerationContext): Any?

	/**
	 * Return the description of this property, or `null` if it is not documented.
	 */
	abstract fun description(context: MetadataGenerationContext): String?

	/**
	 * Return the getter that a nested group should be attributed to, or `null` if the
	 * property has none.
	 */
	protected abstract fun sourceMethod(): String?

	/**
	 * Return whether this property is a nested group of properties.
	 */
	fun isNested(context: MetadataGenerationContext): Boolean {
		val typeDeclaration = type.declaration as? KSClassDeclaration ?: return false
		if (typeDeclaration.classKind == ClassKind.ENUM_CLASS || typeDeclaration.classKind == ClassKind.ENUM_ENTRY) {
			return false
		}
		if (annotatedElements.any { it.hasAnnotation(CONFIGURATION_PROPERTIES) }) {
			return false
		}
		if (annotatedElements.any { it.hasAnnotation(NESTED_CONFIGURATION_PROPERTY) }) {
			return true
		}
		return !isCyclePresent(typeDeclaration, context) && hasSameTopLevelType(typeDeclaration, context)
	}

	/**
	 * Return the metadata for this descriptor, or `null` if it contributes none.
	 */
	fun toItemMetadata(prefix: String, context: MetadataGenerationContext): ItemMetadata? {
		if (isNested(context)) {
			return toGroupMetadata(prefix, context)
		}
		if (isProperty(context)) {
			return toPropertyMetadata(prefix, context)
		}
		return null
	}

	private fun toGroupMetadata(prefix: String, context: MetadataGenerationContext): ItemMetadata {
		val nestedType = type.declaration as KSClassDeclaration
		return ItemMetadata.newGroup(
			ConfigurationMetadata.nestedPrefix(prefix, name),
			context.typeNames.resolve(nestedType),
			context.typeNames.resolve(declaringType),
			sourceMethod(),
		)
	}

	private fun toPropertyMetadata(prefix: String, context: MetadataGenerationContext): ItemMetadata {
		val deprecation = if (context.isDeprecated(*deprecationElements(), declaringType)) {
			context.resolveDeprecation(*deprecationElements())
		}
		else {
			null
		}
		return ItemMetadata.newProperty(
			prefix,
			name,
			context.typeNames.resolve(type),
			context.typeNames.resolve(declaringType),
			null,
			description(context),
			defaultValue(context),
			deprecation,
		)
	}

	private fun deprecationElements(): Array<KSAnnotated> = deprecationElements.toTypedArray()

	/**
	 * Return the name of the JVM getter of this property, mirroring the source method
	 * recorded by the Java annotation processor.
	 */
	protected fun jvmGetterName(): String {
		if (name.length > IS_PREFIX.length && name.startsWith(IS_PREFIX) && !name[IS_PREFIX.length].isLowerCase()) {
			return "$name()"
		}
		return "get" + name.replaceFirstChar(Char::uppercaseChar) + "()"
	}

	private fun isCyclePresent(typeDeclaration: KSClassDeclaration, context: MetadataGenerationContext): Boolean {
		val target = context.typeNames.resolve(typeDeclaration)
		var candidate = declaringType.parentDeclaration
		while (candidate is KSClassDeclaration) {
			if (context.typeNames.resolve(candidate) == target) {
				return true
			}
			candidate = candidate.parentDeclaration
		}
		return false
	}

	private fun hasSameTopLevelType(
		typeDeclaration: KSClassDeclaration,
		context: MetadataGenerationContext,
	): Boolean {
		val target = context.typeNames.resolve(topLevelType(typeDeclaration))
		var candidate: KSClassDeclaration? = declaringType
		while (candidate != null) {
			if (context.typeNames.resolve(topLevelType(candidate)) == target) {
				return true
			}
			candidate = candidate.superClassDeclaration()
		}
		return false
	}

	private fun topLevelType(declaration: KSClassDeclaration): KSClassDeclaration {
		var candidate = declaration
		while (candidate.parentDeclaration is KSClassDeclaration) {
			candidate = candidate.parentDeclaration as KSClassDeclaration
		}
		return candidate
	}

	private fun KSClassDeclaration.superClassDeclaration(): KSClassDeclaration? = superTypes
		.map { it.resolve().declaration }
		.filterIsInstance<KSClassDeclaration>()
		.firstOrNull { it.classKind == ClassKind.CLASS && it.qualifiedName?.asString() != KOTLIN_ANY }

	/**
	 * A [PropertyDescriptor] for a parameter of the constructor used for binding.
	 */
	class ConstructorParameter(
		name: String,
		type: KSType,
		declaringType: KSClassDeclaration,
		private val parameter: KSValueParameter,
		private val property: KSPropertyDeclaration?,
	) : PropertyDescriptor(name, type, declaringType) {

		override val annotatedElements: List<KSAnnotated> =
			listOfNotNull(parameter, property?.getter, property, property?.setter)

		override fun isProperty(context: MetadataGenerationContext): Boolean = !isNested(context)

		override fun defaultValue(context: MetadataGenerationContext): Any? {
			val annotation = annotatedElements.firstNotNullOfOrNull { it.findAnnotation(DEFAULT_VALUE) }
			if (annotation != null) {
				return DefaultValues.fromAnnotation(annotation, type, parameter, context)
			}
			// KSP does not expose parameter default values, see google/ksp#1868
			if (parameter.hasDefault) {
				return null
			}
			return DefaultValues.fromType(type)
		}

		override fun description(context: MetadataGenerationContext): String? = context.documentation(property)
			?: parameter.name?.asString()?.let { context.parameterDocumentation(declaringType, it) }

		override fun sourceMethod(): String? = if (property != null) jvmGetterName() else null

	}

	/**
	 * A [PropertyDescriptor] for a mutable property bound using its getter and setter.
	 */
	class JavaBean(
		name: String,
		type: KSType,
		declaringType: KSClassDeclaration,
		private val property: KSPropertyDeclaration,
		private val factoryMethod: KSFunctionDeclaration? = null,
	) : PropertyDescriptor(name, type, declaringType) {

		override val annotatedElements: List<KSAnnotated> =
			listOfNotNull(property.getter, property, property.setter)

		override val deprecationElements: List<KSAnnotated>
			get() = annotatedElements + listOfNotNull(this.factoryMethod)

		override fun isProperty(context: MetadataGenerationContext): Boolean {
			if (context.isExcluded(type)) {
				return false
			}
			return property.isMutable || context.isCollectionOrMap(type)
		}

		// KSP does not expose property initializers, see google/ksp#1868
		override fun defaultValue(context: MetadataGenerationContext): Any? = null

		override fun description(context: MetadataGenerationContext): String? = context.documentation(property)

		override fun sourceMethod(): String = jvmGetterName()

	}

	private companion object {

		private const val IS_PREFIX = "is"

		private const val KOTLIN_ANY = "kotlin.Any"

	}

}
